File: LeetCode/Two Pointers/Medium/167. Two Sum II - Input Array Is Sorted/main.py
class Solution:
    # using dict
    # def twoSum(self, numbers: List[int], target: int) -> List[int]:
    #     num_dict = {}
    #     for i,num in enumerate(numbers):
    #         val = target - num
    #         if val in num_dict:
    #             return [num_dict[val]+1,i+1]
    #         num_dict[num] = i
    #         print(num_dict)
    #     return []

    # using pointers , faster comparatively
    def twoSum(self, numbers, target: int):
        l, r = 0, len(numbers) - 1
        while r >= l:
            res = numbers[r] + numbers[l]
            if res == target:
                return [l+1, r+1]
            elif res > target:
                r -= 1
            else:
                l += 1


obj = Solution()
print(obj.twoSum([2, 7, 11, 15], 9))

File: LeetCode/Two Pointers/Medium/167. Two Sum II - Input Array Is Sorted/main.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
using namespace std;

class Solution
{
public:
    // vector<int> twoSum(vector<int>& numbers, int target) {
    //     unordered_map<int, int> num;
    //     for(int i = 0;i<numbers.size();i++){
    //         int val = target - numbers[i];
    //         if(num.count(val)){
    //             return {num[val]+1,i+1};
    //         }
    //         num[numbers[i]] = i;
    //     }
    //     return {};
    // }

    // Pointer approch relatively faster
    vector<int> twoSum(vector<int> &numbers, int target)
    {
        int l = 0, r = numbers.size() - 1;
        while (l < r)
        {
            if (target < numbers[l] + numbers[r])
                r--;
            else if (target > numbers[l] + numbers[r])
                l++;
            else
                return {l + 1, r + 1};
        }

        return {};
    }
};

int main()
{
    Solution obj;
    vector<int> v{2, 7, 11, 15};
    vector<int> ans;
    ans = obj.twoSum(v, 9);
    for (auto i = ans.begin(); i != ans.end(); ++i)
        cout << *i << " ";
}